tests/src/tools/ReadyToRun.SuperIlc/Application.cs

```diff
@@ -48,25 +55,67 @@ public class Application
+                    HashSet<string> folders = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+                    modules.Add(_mainExecutable);
+                    modules.Add(runner.GetOutputFileName(_outputFolder, _mainExecutable));
+                    modules.UnionWith(_compilationInputFiles);
+                    modules.UnionWith(_compilationInputFiles.Select(file => runner.GetOutputFileName(_outputFolder, file)));
+                    folders.Add(Path.GetDirectoryName(_mainExecutable));
+                    folders.UnionWith(runner.ReferenceFolders);
+
+                    _execution[(int)runner.Index] = runner.ExecutionProcess(_outputFolder, _mainExecutable, modules, folders, coreRunPath, noEtw);
+                }
+            }
+        }
+
+        public static Application FromDirectory(string inputDirectory, IEnumerable<CompilerRunner> compilerRunners, string outputRoot, bool noEtw, string coreRunPath)
+        {
+            List<string> compilationInputFiles = new List<string>();
+            List<string> passThroughFiles = new List<string>();
+            string mainExecutable = null;
+
+            // Copy unmanaged files (runtime, native dependencies, resources, etc)
+            foreach (string file in Directory.EnumerateFiles(inputDirectory))
+            {
+                bool isManagedAssembly = ComputeManagedAssemblies.IsManaged(file);
+                if (isManagedAssembly)
+                {
+                    compilationInputFiles.Add(file);
+                }
+                else
+                {
+                    passThroughFiles.Add(file);
+                }
+                if (Path.GetExtension(file).Equals(".exe", StringComparison.OrdinalIgnoreCase))
+                {
+                    mainExecutable = file;
+                }
+            }
 
-                    modules.Add(_mainExecutable.ToLower());
-                    modules.Add(runner.GetOutputFileName(_mainExecutable).ToLower());
-                    modules.UnionWith(_compilationInputFiles.Select(file => file.ToLower()));
-                    modules.UnionWith(_compilationInputFiles.Select(file => runner.GetOutputFileName(file).ToLower()));
-                    folders.Add(Path.GetDirectoryName(_mainExecutable).ToLower());
-                    folders.UnionWith(runner.ReferenceFolders.Select(folder => folder.ToLower()));
+            if (compilationInputFiles.Count == 0)
+            {
+                return null;
+            }
 
-                    _execution[(int)runner.Index] = runner.ExecutionProcess(_mainExecutable, modules, folders, coreRunPath);
+            foreach (CompilerRunner runner in compilerRunners)
+            {
+                string runnerOutputPath = runner.GetOutputPath(outputRoot);
+                runnerOutputPath.RecreateDirectory();
+                foreach (string file in passThroughFiles)
+                {
+                    File.Copy(file, Path.Combine(runnerOutputPath, Path.GetFileName(file)));
```

Comment:
We probably want some exception handling around file IO operations so that we can fail a single test and not bring down the whole process if something weird happens. Especially since this is going to be massively parallelized (sp?)

Reply:
I think the ParallelRunner is sufficiently hardened against this so this is mostly about these special steps. For now I added a try block in CompileSubtreeCommand to watch out for failures in identification of a single test assembly; in CompileDirectoryCommand it doesn't make much sense as you're hosed if the application singleton construction fails.